src/source/channel_router.rs

```diff
@@ -0,0 +1,304 @@
+
+/// An input channels to output channels mapping.
+pub type ChannelMap = Vec<ChannelLink>;
+
+/// Function that builds a [`ChannelMixer`] object.
+/// The new `ChannelMixer` will read samples from `input` and will mix and map them according
+/// to `channel_mappings` into its output samples.
+pub fn channel_mixer<I>(
+    input: I,
+    out_channels_count: u16,
+    channel_map: &ChannelMap,
+) -> (ChannelMixer, ChannelMixerSource<I>)
+where
+    I: Source,
+    I::Item: Sample,
+{
+    let (tx, rx) = channel();
+    let controller = ChannelMixer {
+        sender: tx,
+        out_channels_count,
+    };
+    let source = ChannelMixerSource {
+        input,
+        channel_map: vec![],
+        // this will cause the input buffer to fill on first call to next()
+        current_channel: out_channels_count,
+        channel_count: out_channels_count,
+        input_frame: Cell::new(vec![]),
+        // I::Item::zero_value() zero value is not 0 for some sample types,
+        // so have to use an option.
+        output_frame: vec![None; out_channels_count.into()],
+        receiver: rx,
+    };
+    // TODO Return an error here? Requires to change API. Alternatively,
+    //      map can be set separately, or this can panic.
+    controller
+        .set_map(channel_map)
+        .expect("set channel mixer map");
+
+    (controller, source)
+}
+
+/// `ChannelRouterController::map()` returns this error if the router source has been dropped.
+#[derive(Debug, Eq, PartialEq)]
+pub enum ChannelMixerError {
+    ConfigError,
+    SendError,
+}
+
+impl fmt::Display for ChannelMixerError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "<ChannelMixerError>")
+    }
+}
+
+impl Error for ChannelMixerError {}
+
+/// A controller type that sends gain updates to a corresponding [`ChannelMixerSource`].
+#[derive(Debug, Clone)]
+pub struct ChannelMixer {
+    sender: Sender<ChannelMap>,
+    out_channels_count: ChannelCount,
+}
+
+impl ChannelMixer {
+    /// Set or update the gain setting for a channel mapping.
+    pub fn set_map(&self, channel_map: &ChannelMap) -> Result<(), ChannelMixerError> {
+        let mut new_map = channel_map.clone();
+        self.prepare_map(&mut new_map)?;
+        self.sender
+            .send(new_map)
+            .map_err(|_| ChannelMixerError::SendError)
+    }
+
+    fn prepare_map(&self, new_channel_map: &mut ChannelMap) -> Result<(), ChannelMixerError> {
+        if !new_channel_map
+            .iter()
+            .all(|(_from, to, _gain)| to < &self.out_channels_count)
+        {
+            return Err(ChannelMixerError::ConfigError);
+        }
+        new_channel_map.retain(|(_from, _to, gain)| *gain != 0.0);
+        new_channel_map.sort_by(|a, b| a.0.cmp(&b.0));
+        Ok(())
+    }
+}
+
+/// A source for extracting, reordering, mixing and duplicating audio between
+/// channels.
+// #[derive(Debug)] // TODO Reimplement debug? A Cell is not Debug.
```

Comment:
might want to implement it even without Cell. The derived implementation will print input_frame, output_frame and receiver. Those might not be very useful for debugging but will spam up the terminal.